---
title: "Icm.Microarray.RNASeq"
author: "Enrique ESCOBAR"
date: "20 Feb 2017"
output: html_document
---

# R Project Setup

A file Script.R that sets paths.

```{r project setup, include=FALSE, echo=TRUE}
# modify this path to you local repository
# Windows machines have a environment variable SystemDrive=C:
# I use my own for working directory as DataDrive & REPO_DIR
localRepository <- Sys.getenv("DataDrive");
localRepository <- paste0(localRepository, "/", Sys.getenv("REPO_DIR"));
write(paste0(c("Repository ...\t", localRepository), sep = "", collapse = ""), stdout());
# solution is the folder containing the project
projectSolution <- "Kinito.Revolution";
write(paste0(c("Solution .....\t", projectSolution), sep = "", collapse = ""), stdout());
# project is the project's name
projectName <- "Icm.Microarray.RNASeq";
write(paste0(c("Project ......\t", projectName), sep = "", collapse = ""), stdout());
# namespace
projectNamespace <- "";
projectNamespace <- if(projectSolution == projectName) projectSolution else paste0(projectSolution, ".", projectName);
write(paste0(c("Namespace ....\t", projectNamespace), sep = "", collapse = ""), stdout());
# common
projectCommon <- "RCommon";
write(paste0(c("Common .......\t", projectCommon), sep = "", collapse = ""), stdout());
# path
projectPath <- "";
projectPath <- if(projectSolution == projectName) projectSolution else paste0(projectSolution, "/", projectName);
write(paste0(c("Path .........\t", projectPath), sep = "", collapse = ""), stdout());
projectPath <- paste0(localRepository, "/", projectPath);
write(paste0(c("Path old .....\t", getwd()), sep = "", collapse = ""), stderr());
write(paste0(c("Path new .....\t", projectPath), sep = "", collapse = ""), stdout());
setwd(projectPath);
# clean
rm(localRepository);
# options
options(scipen = 100);
# session info
sessionInfo();
# attached
search();
# Create a listing of all objects in the "global environment".
ls();
```

# Data packaging

## Reading

```{r data-packaging-reading}
source(paste0("Lib/", projectName, ".Util.R"));
fileList <- c("Data/GSM1545535_10_6_5_11.txt",
           "Data/GSM1545536_9_6_5_11.txt",
           "Data/GSM1545538_purep53.txt",
           "Data/GSM1545539_JMS8-2.txt",
           "Data/GSM1545540_JMS8-3.txt",
           "Data/GSM1545541_JMS8-4.txt",
           "Data/GSM1545542_JMS8-5.txt",
           "Data/GSM1545544_JMS9-P7c.txt",
           "Data/GSM1545545_JMS9-P8c.txt");
```

# GeneNames *col1* & GeneCount *col3*

```{r data-packaging-genenames}
df <- TxtFileToDataFrame(fileList[1]);
head(df);
geneColumns <- which(colnames(df) == "EntrezID" | colnames(df) == "Count");
geneColumns;
geneKeyType <- toupper(colnames(df)[1]);
geneKeyType;
rm(df);
source(paste0("Lib/", "edgeR.Util.R"));
x <- GeneNameAndCountDGEList(fileList, geneColumns);
dim(x);
```

## Organising sample information

```{r data-packaging-info}
### shrink colnames remove from start to first underscore
colnames(x);
samplenames <- substring(colnames(x), 17, nchar(colnames(x))) ; #12->17
samplenames;
colnames(x) <- samplenames;
# col group & norm.factors preset to 1
x$samples;
### group factor foreach-9
group <- as.factor(c("LP", "ML", "Basal", "Basal", "ML", "LP", "Basal", "ML", "LP")) ;
x$samples$group <- group;
x$samples;
#rm(group);
### lane factor foreach-9
lane <- as.factor(rep(c("L004", "L006", "L008"), c(3, 4, 2))) ;
x$samples$lane <- lane;
x$samples;
```

## Gene annotations

```{r data-packaging-annotations-lib, include=FALSE, echo=FALSE}
#source("https://bioconductor.org/biocLite.R");
#biocLite("Mus.musculus");
#library(Mus.musculus);
source(paste0("Lib/", "Mus.musculus.Util.R"));
```

```{r data-packaging-annotations, echo=TRUE}
# loads library
# AnnotationDbi, stats4, BiocGenerics, parallel, Biobase, IRanges, S4Vectors
# OrganismDbi, GenomicFeatures, GenomeInfoDb, GenomicRanges, GO.db, org.Mm.eg.db,
# TxDb.Mmusculus.UCSC.mm10.knownGene
head(rownames(x), 10);
geneIdKeys <- rownames(x) ;
### foearch rowname select cols in ENTREZID
geneColumns <- c("SYMBOL", "TXCHROM");
#genes <- AnnotationDbi::select(Mus.musculus, keys = geneIdKeys, columns = geneColumns, keytype = geneKeyType);
# select() returned 1:many mapping between keys and columns
genes <- EntrezIdSymbolChromosome(geneList = geneIdKeys, geneColumns = geneColumns, geneKeyType = geneKeyType);
dim(genes);
colnames(genes);
head(genes);
### rm duplicates
genes <- genes[!duplicated(genes$ENTREZID),];
dim(genes);
colnames(genes);
head(genes);
### match genes
x$genes <- genes;
x;
```

# Data pre-processing

## Transformations from the raw-scale

```{r data-pre-processing-transformations}
#simpleCountsPerMillion <- edgeR::cpm(x) ;
simpleCountsPerMillion <- CountsPerMillion(x);
#logCountsPerMillion <- edgeR::cpm(x, log = TRUE);
logCountsPerMillion <- CountsPerMillion(x, TRUE);
## Removing genes that are lowly expressed
table(rowSums(x$counts == 0) == 9);
keep.exprs <- rowSums(simpleCountsPerMillion > 1) >= 3;
x <- x[keep.exprs, , keep.lib.sizes = FALSE];
dim(x);
#library(RColorBrewer);
source(paste0("Lib/", "RColorBrewer.Util.R"));
sampleCount <- ncol(x);
#pairedColors <- RColorBrewer::brewer.pal(sampleCount, "Paired");
pairedColors <- ColorPalette(sampleCount, "Paired");
par(mfrow = c(1, 2));
plot(density(logCountsPerMillion[, 1]), col = pairedColors[1], lwd = 2, ylim = c(0, 0.21), las = 2, main = "", xlab = "");
title(main = "A. Raw data", xlab = "Log-cpm");
abline(v = 0, lty = 3);
for (i in 2:sampleCount){
  den <- density(logCountsPerMillion[, i]);
  lines(den$x, den$y, col = pairedColors[i], lwd = 2);
}
legend("topright", samplenames, text.col = pairedColors, bty = "n");

#logCountsPerMillion <- edgeR::cpm(x, log = TRUE);
logCountsPerMillion <- CountsPerMillion(x, TRUE);
plot(density(logCountsPerMillion[, 1]), col = pairedColors[1], lwd = 2, ylim = c(0, 0.21), las = 2, main = "", xlab = "");
title(main = "B. Filtered data", xlab = "Log-cpm");
abline(v = 0, lty = 3);
for (i in 2:sampleCount){
  den <- density(logCountsPerMillion[, i]);
  lines(den$x, den$y, col = pairedColors[i], lwd = 2);
}
legend("topright", samplenames, text.col = pairedColors, bty = "n");
```

## Normalizing gene expression distributions

```{r data-normalization, echo=TRUE}
#x <- edgeR::calcNormFactors(x, method = "TMM");
x <- CalculateNormalizationFactors(x, "TMM");
x$samples$norm.factors;
x2 <- x
x2$samples$norm.factors <- 1;
x2$counts[, 1] <- ceiling(x2$counts[, 1]*0.05);
x2$counts[, 2] <- x2$counts[, 2]*5;
par(mfrow = c(1, 2));
#logCountsPerMillion <- edgeR::cpm(x2, log=TRUE);
logCountsPerMillion <- CountsPerMillion(x2, TRUE);
boxplot(logCountsPerMillion, las = 2, col = pairedColors, main ="");
title(main = "A. Example: Unnormalised data", ylab = "Log-cpm");

#x2 <- edgeR::calcNormFactors(x2);
x2 <- CalculateNormalizationFactors(x2);
x2$samples$norm.factors;
#logCountsPerMillion <- edgeR::cpm(x2, log=TRUE);
logCountsPerMillion <- CountsPerMillion(x2, TRUE);
boxplot(logCountsPerMillion, las = 2, col = pairedColors, main = "");
title(main = "B. Example: Normalised data", ylab = "Log-cpm");
```

## Unsupervised clustering of samples

# Differential expression analysis

## Creating a design matrix and contrasts

## Removing heteroscedascity from count data

## Fitting linear models for comparisons of interest

## Examining the number of DE genes

## Examining individual DE genes from top to bottom

## Useful graphical representations of differential expression results BasalvsLP

# Gene set testing with camera

